go-interface-tutorial/main_test.go

```diff
@@ -101,3 +107,127 @@ func addBookWithTitle(title string) {
 
 	p.sdb.Query("INSERT INTO books VALUES ($1)", title)
 }
+
+func TestHandlerWithNoQuery(t *testing.T) {
+	connStr := "host=localhost port=5400 user=docker password=docker sslmode=disable"
+	db, err := sql.Open("postgres", connStr)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer db.Close()
+
+	req, err := http.NewRequest("GET", "/", nil)
+	if err != nil {
+		return
+	}
+	w := httptest.NewRecorder()
+	api := API{
+		repository: PostgresRepository{
+			sdb: &ShopDB{db},
+		},
+	}
+	api.searchHandler(w, req)
+
+	resp := w.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+	fmt.Println(string(body))
+
+	if resp.Status != "400 Bad Request" {
+		t.Error("Supposed to be no query")
+	}
+}
+
+func TestHandlerWithNoResults(t *testing.T) {
+	connStr := "host=localhost port=5400 user=docker password=docker sslmode=disable"
+	db, err := sql.Open("postgres", connStr)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer db.Close()
+
+	req, err := http.NewRequest("GET", "/?search=testing", nil)
+	if err != nil {
+		return
+	}
+	w := httptest.NewRecorder()
+	api := API{
+		repository: PostgresRepository{
+			sdb: &ShopDB{db},
+		},
+	}
+	api.searchHandler(w, req)
+
+	resp := w.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+	fmt.Println(string(body))
+
+	type Result struct {
+		Results []string `json:"results"`
+		Count   int      `json:"count"`
+	}
+
+	var result Result
+
+	test := json.Unmarshal(body, &result)
+	if test != nil {
+		t.Error("Unmarshal error")
+	}
+
+	if result.Count != 0 {
+		t.Error("Expected 0 results. Results not 0.")
+	}
+}
+
+func TestHandlerWithSeveralResults(t *testing.T) {
+	connStr := "host=localhost port=5400 user=docker password=docker sslmode=disable"
+	db, err := sql.Open("postgres", connStr)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer db.Close()
+
+	req, err := http.NewRequest("GET", "/?search=testing", nil)
+	if err != nil {
+		return
+	}
+	w := httptest.NewRecorder()
+	api := API{
+		repository: PostgresRepository{
+			sdb: &ShopDB{db},
+		},
+	}
+	api.repository.sdb.Query("INSERT INTO books VALUES ($1)", "testing")
+	api.repository.sdb.Query("INSERT INTO books VALUES ($1)", "testingcute")
+	api.repository.sdb.Query("INSERT INTO books VALUES ($1)", "testingpinch")
+	api.searchHandler(w, req)
+	api.repository.sdb.Query("DELETE FROM books WHERE title=$1", "testing")
+	api.repository.sdb.Query("DELETE FROM books WHERE title=$1", "testingcute")
+	api.repository.sdb.Query("DELETE FROM books WHERE title=$1", "testingpinch")
+
+	resp := w.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+	fmt.Println(string(body))
+
+	type Result struct {
+		Results []string `json:"results"`
+		Count   int      `json:"count"`
+	}
+
+	var result Result
+
+	test := json.Unmarshal(body, &result)
+	if test != nil {
+		t.Error("Unmarshal error")
+	}
+
+	if result.Count != 3 {
+		t.Error("Expected 3 results. Results not 3.")
+	}
+
+	if strings.Join(result.Results, ",") != "testing,testingcute,testingpinch" {
```

Comment:
I would sort the results before you do this otherwise, you may get failures in the future if the sort order changes.